hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -57,6 +58,54 @@ isStackCradle = (`elem` ["stack", "Cabal-Helper-Stack", "Cabal-Helper-Stack-None
   . BIOS.actionName
   . BIOS.cradleOptsProg
 
+  -- | Check if the given cradle is a cabal cradle.
+-- This might be used to determine the GHC version to use on the project.
+-- If it is a stack-cradle, we have to use `stack path --compiler-exe`
+-- otherwise we may ask `ghc` directly what version it is.
+isCabalCradle :: Cradle -> Bool
+isCabalCradle =
+  (`elem`
+    ["cabal"
+    , "Cabal-Helper-Cabal-V1"
+    , "Cabal-Helper-Cabal-V2"
+    , "Cabal-Helper-Cabal-V1-Dir"
+    , "Cabal-Helper-Cabal-V2-Dir"
+    , "Cabal-Helper-Cabal-None"
+    ]
+  )
+  . BIOS.actionName
+  . BIOS.cradleOptsProg
+
+
+getProjectGhcPath :: Cradle -> IO (Maybe FilePath)
+getProjectGhcPath crdl = do
+  isStackInstalled <- isJust <$> findExecutable "stack"
+  isCabalInstalled <- isJust <$> findExecutable "cabal"
+  if isStackCradle crdl && isStackInstalled
+    then
+      catch (Just <$> tryCommand "stack path --compiler-exe") $ \(_ :: IOException) ->
+        return Nothing
+    else if isCabalCradle crdl && isCabalInstalled then do
+      Just ghcCabalVersion <- catch (Just <$> tryCommand "cabal v2-exec ghc -- --numeric-version") $ \(_ ::IOException) ->
+        return Nothing
+      findExecutable ("ghc-" ++ ghcCabalVersion)
+    else
+      findExecutable "ghc"
+
+tryCommand :: String -> IO String
+tryCommand cmd =
+  T.unpack . T.strip .T.pack <$> readCreateProcess (shell cmd) ""
+
+getProjectGhcLibDir :: Cradle -> IO (Maybe FilePath)
+getProjectGhcLibDir crdl = do
+  mGhcPath <- getProjectGhcPath crdl
+  case mGhcPath of
+    Nothing -> return Nothing
+    Just ghcPath -> catch (Just <$> tryCommand (ghcPath ++ " --print-libdir")) $ \(_ :: IOException) -> return Nothing
```

Comment:
Catching and discarding the exception is probably a bad thing.

At the very least we should log something on failure.

Is the the exception that would be thrown if GHC is not installed?  Do we report it gracefully elsewhere if that is the case?

Reply:
No, not yet. Not sure how to properly report it, since it is executed before the server is actually started, e.g. When we start the GHC thread

Reply:
Perhaps we should re-look at this startup sequencing, and launch the various processes once the initialize message comes in, while processing it.

That way we can determine the cradle according to the passed-in project root, and we have a way to report errors.  The spec says we are not to process anything before initialization is complete anyway, so it *is* an appropriate place to do it.

cc @bubba @mpickering

Reply:
Yes, I think the LSP server should be bootstrapped before the dispatcher. And while we're at it, we should provide an easier way to log/show these messages via LSP. Presumably some sort of transformer stack hierarchy:

```
IO a < LspM a < IdeM a < IdeGhcM a
```

Reply:
Would that mean to execute `onStartup` https://github.com/haskell/haskell-ide-engine/blob/master/src/Haskell/Ide/Engine/Server.hs#L125 in `NotInitialized`?

Reply:
No, `onStartup` is the one that is called when the initialize request comes in. At this point haskell-lsp has comms established with the client, and has received a capability list from it, as well as the root directory.

So we can find the libdir and launch the dispatchers from here, and if there is a problem send back a message which will be displayed on the client.

Reply:
See https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#initialize

Reply:
And perhaps we should rename `onStartup` to `onInitialize`, and make it clear where in the server life cycle it operates.

Reply:
Oh, so that is actuallly fine as-is? Just need to refactor when the cradle is loaded?

Reply:
yes

Reply:
And this change should also resolve https://github.com/alanz/vscode-hie-server/issues/164

Reply:
Not entirely. For hie-bios multi-cradles, also #1490 needs to be resolved.